### ISSUE TITLE
make fuzzy symbol searching smarter

### ISSUE THREAD

**Issue by synkarius** [synkarius/caster#49]
https://github.com/synkarius/caster/blob/master/caster/lib/pita/selector.py#L11

My original shot at a fuzzy string matching algorithm sucks. It needs to be better.

**Comment by synkarius** [comment]
NEEDED: test data to try out different algorithms and variations on;

Please submit symbols from any codebase you like. If you could format as follows, that'd be great too:
[spoken word(s), actual answer, choice1, choice2, choice3, etc.]

Examples:
["issue maturity date", "ismtdt", "32d22as", "i", "date", "ismat"]
["certificate cancel amount", "ccnamt", "ccd", "cancel", "certs"]
["stock ticker", "stktck", "stock", "stTick", "sstkks", "tick123"]

**Comment by chilimangoes** [comment]
Per #122, added logging to the FuzzyMatchAction to facilitate data collection in the above format. Logging is disabled by default and can be opted in by setting ["pita"]["enable_logging"] to true in `settings.json`.

**Comment by chilimangoes** [comment]
I just remembered I have an an algorithm I wrote a couple years ago to match YouTube video titles to theTVDB episode titles. The use case is a little different, but it might work for this. I'll dig it up and port it to python so we can test it.

**Comment by synkarius** [comment]
Excellent. The more we try the better. It could turn out that some algorithms fit some codebases better than others, though in the (few admittedly limited) tests I've done so far, difflib has (unsurprisingly) done consistently better than my homemade matching algorithm.

**Comment by chilimangoes** [comment]
Here are some brainstorming notes on how we might take a machine learning approach to this. We could use a recurrent neural net (RNN) to map strings to vectors and then use those vectors to determine similarity, perhaps by calculating euclidean distance between vectors or feeding the vectors into a second model that outputs a similarity score. Although not the same thing as what we're doing, I think there are some similarities to training [neural language and translation models](https://github.com/kjw0612/awesome-rnn/blob/master/README.md#machine-translation) to map words in different languages with a similar meaning to similar locations in a vector space.

One potential benefit of this, from a performance standpoint, is the process of mapping symbols to vectors could be done as part of the initial scanning process and stored in a hash table for quick retrieval. I don't know whether this would make a big difference, since we would still need to compute the vector for the spoken phrase at runtime, but it might help.

A relatively small amount of training data would be a problem, so here are some ideas for how we might be able to work around that limitation:

* Use an auto encoder on a large corpus of random English phrases and symbols collected from various code bases to produce an RNN that can encode text char by char into a vector. Then using that as a starting point, train to minimize distance between spoken phrases and matching symbols, while maximizing the distance between non-matching pairs.

* If we can find an algorithm or combination of algorithms that work well but are to slow, we could use the techniques that Geoff Hinton discusses in [Dark Knowledge](https://www.google.com/search?q=geoff+hinton+dark+knowledge) to distill the "knowledge" embedded in these less performant models/algorithms into a smaller, more performant RNN. For example, you mentioned Levenstein and difflib. Training an RNN on random combinations of phrases, code symbols, and even random character strings and using Levenstein as the objective function would eventually teach the RNN to produce the same (or very close) output as Levenstein, but possibly be much faster. And then we could try to further improve the performance of the resulting model by doing additional training using our real training data.

**Comment by synkarius** [comment]
I love this to death, but I know nothing about RNNs. So, the following questions may be more a reflection of my ignorance than legitimate questions. That said, please bear with me as I attempt to understand.

- If we scan a ton of code bases, do we then have to go through the scanned symbols and pronounce them? Is that what the training consists of? 
- To avoid the slowness problem, wouldn't it be easier to just make a C++ DLL which does the job that we want done quickly? For instance, if Levenshtein turned out to be the best.
- When you say "vector", I hear "array of characters", but that's probably not what you mean. What do you mean?

**Comment by chilimangoes** [comment]
No problem. I didn't want to go into excess detail on things you already knew. But I don't mind explaining anything you have questions about. Also, I've studied RNNs quite a bit but don't have any practical experience with them, so this will be a learning experience for me too. And if you're interested in ML, I can't recommend enough Andrew Ng's [ML course](https://www.coursera.org/learn/machine-learning) on Courera. It's a pretty big time investment to actually complete the course and all the homework, but I took it a few years ago and it was one of my favorite courses on any subject ever. The lectures are also available to view if you just want a (relatively) quick primer on some of the subjects.

To your questions:

1) By training, I mean the automated learning process that the algorithm goes through to adjust the parameters of the model (a neutral network in this case) to give good answers for the examples in our training data set. For example, we would feed the algorithm example phrase/symbol pairs from the data we collect in the format you requested above, along with an answer key for whether each pair represents a match or not. The network then calculates what it thinks the probability of a match is, compares its answer to the ground truth answer we gave it, and adjusts the parameters of the model to make its answer closer to the truth. Rinse and repeat for each example in our training data set. Also, generally speaking, the more training data you have, the better. Tens of thousands or millions of examples would be great, but not realistic for us to compile, so I threw out a couple ideas for how we might mitigate this issue by essentially pre-training the network on a large volume of unlabeled data (first option) or automatically generated data (second option). If you want me to go into more detail on either of those options, just ask.

2) The bulk of processing work in most machine learning algorithms is linear algebra - manipulating and performing arithmetic on vectors and matrices of real numbers. There are a number of linear algebra libraries out there, written in C, with python wrappers. When I talk about training an RNN to produce similar output to say Levenstein, I mean we would encode the string as a series of real numbers, and then the RNN would essentially learn a mathematical transformation on those input numbers to produce an output number close to the number produced by Levenstein for that same input. Because it's all just basic number crunching, it could *potentially* be faster than a manually-coded algorithm that's doing lots of string and character manipulation - but we won't know until we try.

3) As you probably guessed from (2), I mean a linear algebra vector. In python terms, a vector used in ML would typically be an array of real numbers and a matrix would either be an array of arrays of real numbers or a two-dimensional array of real numbers.

**Comment by synkarius** [comment]
Thanks for clarifying. After reading over this a few times, I'm wondering too, if this approach might handle underscores and dashes in variable names better than distance algorithms would, since spoken phrases produced by Dragon will seldom have those characters in them.

I think this is definitely worth a shot at least, though most of the work will be on you, since the extent of my machine learning knowledge is writing a simple genetic algorithm once. I think the whole subject is fascinating too, but as much as I'd like to, I don't have time to take an online course at the moment.

After trying a few more distance algorithms (on actual data now that #127 is done), I may pick one of them, close this issue, and reopen a similar issue and assign you to it, just so I can officially close out the "v6" issues. This sounds like it could be quite a bit of work.

**Comment by chilimangoes** [comment]
I understand being too busy. If I hadn't taken the course when I did, I definitely wouldn't have time for it now.

I think you're right that an ML algorithm should handle things like dashes and underscores. Its whole purpose in life is to learn which inputs are relevant and which aren't. In fact, those characters might even help it in the same way they help us, by marking word boundaries. Which makes me wonder if this might be a good addition to our manual similarity algorithms. It seems like the first letter of each word should be *far* more likely to exist in the resulting abbreviation or acronym. What if we take the consecutive but not-necessarily-contiguous letters idea and give an extra boost to letters at the beginning of words in the spoken phrase? A further boost could be given if the letter is found at a word boundary in the symbol - like after a dash or underscore, or a capitalized letter in a camel/Pascal case symbol.

We could also try to break up abbreviated symbols into words by replacing any non-alphanumeric chars with single spaces and breaking Pascal/camel case symbols where there's a case change. Replacing non-alphanumeric chars with spaces was something I did to clean and normalize inputs to my TV title matching script and it improved my results somewhat.

**Comment by chilimangoes** [comment]
Another idea, more regarding usability than algorithm performance (and maybe this warrants being a separate issue). What about caching the last N symbol matches (possibly even turning them into temporary commands). So for example, I say "symbol match test function" and caster presents a list of matches. I say "numb three" and caster outputs "testfunc" and creates a temporary command with the spec "symbol match test function" (or maybe just "symbol test function").

**Comment by synkarius** [comment]
I agree with your intuition about naming temporary commands so as to not overlap with the matching command's Dictation extra. Better still would be to use some other word entirely, like "memory" (though perhaps something shorter), because a variable name could start with "match". Also, in regard to doing that, there is a bit of machinery already in place to assist: check out `SelfModifyingRule`. It is a kind of MergeRule which reloads itself and re-merges all the affected CCR grammars when it is changed. (Of course, there's no need for that if the temporary commands aren't CCR commands.) The present `SelfModifyingRule`s are the history and alias commands, and NodeRules, each of which needs to modify itself in some way. History and Vanilla Alias are non-CCR. NodeRule can be CCR or not. Chain Alias is always CCR.

**Comment by chilimangoes** [comment]
Good point about variable names potentially starting with "match". I don't know about you, but I have enough other things on my plate right now that I don't know when I might get around to doing this. I'm going to open up a new issue so the idea doesn't fall through the cracks.

**Comment by synkarius** [comment]
I noticed you're a bit loaded up. No pressure. It's not like we have deadlines here. :)

**Comment by synkarius** [comment]
#154 
Symbol Match was cut; this issue is now irrelevant.
